### PR TITLE
[xcpmd] Report aggregate stats for multi-battery systems, and other small improvements

### PR DESCRIPTION
**Pull request by jctemkin**
Rolling a handful of small, orthogonal changes into one PR seemed like it'd make testing more convenient, but I can break it apart if desired.

OXT-368

Relies on [IDL PR11](https://github.com/OpenXT/idl/pull/11).

### CODE REVIEW

File: xcpmd/src/battery.c [hunk 1]
```diff
@@ -100,13 +304,28 @@ int get_battery_percentage(unsigned int battery_index) {
...
+    if (info->last_full_capacity != 0) {
+        capacity = info->last_full_capacity;
+    }
+    else {
+        capacity = info->design_capacity;
+    }
+
     //Avoid dividing by zero
-    if (info->last_full_capacity == 0)
+    if (capacity == 0) {
```
Comment (rneilturner) [inline xcpmd/src/battery.c:317]: This is no longer avoiding the divide by zero.

File: xcpmd/src/battery.c [hunk 2]
```diff
@@ -64,23 +67,223 @@ int get_current_battery_level(void) {
...
+            }
+            if (i < num_battery_structs_allocd) {
+                /* Yes! */
+                /* If the other battery is charging, we're pending charge */
+                if (last_status[i].state & 0x2)
+                    return BATT_PENDING_CHARGE;
+                /* If the other battery is discharging, we're pending discharge */
+                else if (last_status[i].state & 0x1)
+                    return BATT_PENDING_DISCHARGE;
+            } else
```
Comment (rneilturner) [inline xcpmd/src/battery.c]: Redundant, remove else condition and let the unknown condition fall through.

File: xcpmd/src/battery.c [hunk 3]
```diff
@@ -64,23 +67,223 @@ int get_current_battery_level(void) {
...
+
+
+//Returns whether the system as a whole is charging, discharging, full, or empty.
+int get_system_charge_state(void) {
+
+    int percentage = get_overall_battery_percentage();
+    long charge_rate = get_total_charge_rate();
+
+    if (get_ac_adapter_status() == ON_AC) {
+        if (percentage > 98 || charge_rate == 0) {
```
Comment (rneilturner) [inline xcpmd/src/battery.c]: Above we make the case that percentage full is 90%, here we determine full to be 98%. Why the difference?

File: xcpmd/src/battery.c [hunk 4]
```diff
@@ -64,23 +67,223 @@ int get_current_battery_level(void) {
...
+            return BATT_FULL;
+        }
+        else {
+            return BATT_CHARGING;
+        }
+    }
+    else {
+        if (percentage < 5) {
+            return BATT_EMPTY;
```
Comment (rneilturner) [inline xcpmd/src/battery.c:131]: Again, above we make the case for 10% battery empty, here 5%, why?

File: xcpmd/src/battery.c (hunk 1 above)
Comment (jctemkin) [inline xcpmd/src/battery.c:317]: Capacity should be the denominator in line 320; I'll submit a fix.

File: xcpmd/src/battery.c (hunk 2 above)
Comment (jctemkin) [inline xcpmd/src/battery.c]: This was moved from xcpmd_dbus_server.c, and I was trying not to modify code that wasn't mine any more than strictly necessary, but I can remove that if you prefer.

File: xcpmd/src/battery.c (hunk 3 above)
Comment (jctemkin) [inline xcpmd/src/battery.c]: The idea was knowing the charge rate gives us a much greater degree of certainty that the batteries really are full. The percentage was thrown in as a fallback for systems that still report a trickle of battery current even when things are full. That aside, you're right in that there should be consistency--I'll correct this.

File: xcpmd/src/battery.c [hunk 5]
```diff
@@ -64,23 +67,223 @@ int get_current_battery_level(void) {
...
+
+
+//Returns the overall energy flowing in/out of all batteries per hour.
+//Positive return values signify that the batteries are charging; negative
+//values indicate the batteries are discharging.
+static long get_total_charge_rate(void) {
+
+    unsigned int i;
+    int state;
+    unsigned long rate, voltage, total_charge_rate;
```
Comment (rossphilipson) [inline xcpmd/src/battery.c]: voltage is unused
it seems total_charge_rate should be a signed type (long in this case)

File: xcpmd/src/battery.c [hunk 6]
```diff
@@ -64,23 +67,223 @@ int get_current_battery_level(void) {
...
+
+    return total_charge_rate;
+}
+
+
+//Returns the total charge of all batteries in the system.
+static unsigned long get_total_charge(void) {
+
+    unsigned int i;
+    unsigned long charge, voltage;
```
Comment (rossphilipson) [inline xcpmd/src/battery.c]: voltage unused

File: xcpmd/src/battery.c [hunk 7]
```diff
@@ -64,23 +67,223 @@ int get_current_battery_level(void) {
...
+
+    return total_charge;
+}
+
+
+//Returns the sum of the maximum charge of all batteries in the system.
+static unsigned long get_total_max_charge(void) {
+
+    unsigned int i;
+    unsigned long charge, voltage;
```
Comment (rossphilipson) [inline xcpmd/src/battery.c]: voltage unused

File: xcpmd/src/battery.c [hunk 8]
```diff
@@ -64,23 +67,223 @@ int get_current_battery_level(void) {
...
+}
+
+
+//Returns an estimate of the time to fully charge the system, in seconds, or
+//0 if the system isn't charging.
+unsigned int time_to_full(void) {
+
+    long charge_rate;
+    unsigned long max_charge, current_charge;
+    int charge_time, i;
```
Comment (rossphilipson) [inline xcpmd/src/battery.c]: unused i

**Comment by rossphilipson** [comment]
There are a lot of unused local variables. I think for now you should remove the obvious ones. Then later to a commit building with -Werror and clean up xcpmd in general. This should probably be a ticket.

File: xcpmd/src/xcpmd-dbus-server.c [hunk 9]
```diff
@@ -440,20 +437,7 @@ gboolean xcpmd_battery_percentage(XcpmdObject *this, guint IN_bat_n, guint *OUT_
...
-        juice_when_full = last_info[IN_bat_n].design_capacity;
-
-    /* Let's not divide by 0 */
-    if (juice_when_full == 0) {
-        g_set_error(error, DBUS_GERROR, DBUS_GERROR_FAILED, "Unhappy battery: %d", IN_bat_n);
-        return FALSE;
-    }
-
-    *OUT_percentage = juice_left * 100 / juice_when_full;
+    *OUT_percentage = get_battery_percentage(IN_bat_n);
```
Comment (rossphilipson) [inline xcpmd/src/xcpmd-dbus-server.c:440]: A whole bunch of code is removed here and replaced with a call to get_battery_percentage(). I would expect that function to be defined in this PR. I guess I am confused about this.

File: xcpmd/src/xcpmd-dbus-server.c [hunk 10]
```diff
@@ -476,64 +460,61 @@ gboolean xcpmd_battery_is_present(XcpmdObject *this, guint IN_bat_n, gboolean *O
...
-                /* If the other battery is charging, we're pending charge */
-                if (last_status[i].state & 0x2)
-                    *OUT_state = 5;
-                /* If the other battery is discharging, we're pending discharge */
-                else if (last_status[i].state & 0x1)
-                    *OUT_state = 6;
-            } else
-                /* We tried everything, the state is unknown... */
-                *OUT_state = 0;
-        }
```
Comment (rossphilipson) [inline xcpmd/src/xcpmd-dbus-server.c:531]: I do understand why you moved this to its own function.

File: xcpmd/src/vm-actions-module.c [hunk 11]
```diff
@@ -670,3 +688,38 @@ void shutdown_vpnvm_dependencies_of_vm_by_uuid (struct arg_node * args) {
...
+    char ** paths;
+    int num_vms, i;
+
+    //Gather a fresh list of VMs
+    populate_vm_identifier_table();
+
+    //Clone their paths, since shutdown_dependencies_of_vm can free the global
+    //VM identifier table out from under us.
+    num_vms = vm_identifier_table->num_entries;
+    paths = (char **)malloc(num_vms * sizeof(char));
```
Comment (rossphilipson) [inline xcpmd/src/vm-actions-module.c]: Shouldn't this be sizeof(char*)? That looks like a buffer overrun....

File: xcpmd/src/xcpmd-dbus-server.c (hunk 9 above)
Comment (jctemkin) [inline xcpmd/src/xcpmd-dbus-server.c:440]: It was added to battery.c in an [earlier commit](https://github.com/OpenXT/xctools/commit/aca4a564ceb08adae059eb69a7dab5a43253b94b) with the intention of being used in situations like this, but the code in xcpmd-dbus-server wasn't pulled out at that time. This change should probably have belonged to that commit.

File: xcpmd/src/vm-actions-module.c (hunk 11 above)
Comment (jctemkin) [inline xcpmd/src/vm-actions-module.c]: You're right; thanks for catching that.

File: xcpmd/src/xcpmd-dbus-server.c (hunk 10 above)
Comment (jctemkin) [inline xcpmd/src/xcpmd-dbus-server.c:531]: The function get_total_charge_rate() in battery.c needs to know the current charge state of each battery individually.

**Comment by jctemkin** [comment]
I hadn't realized that XCPMD's makefile had -Wno-unused; disabling that shows a lot of places where things can be cleaned up, both with respect to this commit and otherwise. A [ticket](https://openxt.atlassian.net/browse/OXT-441) has been opened.

**Comment by rossphilipson** [comment]
Tested, reviewed and updated ticket. Looks good, will merge soon.

**Pull request merged**